### PR TITLE
[core][state] Task events backend - port profile events and turn on task backend [4/n] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
**Previous PRs:**
 - https://github.com/ray-project/ray/pull/30829: 
 - https://github.com/ray-project/ray/pull/30953: 
 - https://github.com/ray-project/ray/pull/30867: 
 - https://github.com/ray-project/ray/pull/30979: 
 - https://github.com/ray-project/ray/pull/30934

**In this PR:** 
- New task backend on by default.
- Changed the profiling event's generation in `profiling.cc`
  - Removed `Profiler` to use TaskEventsBuffer directly when emitting `ProfileEvent` 
- Changed front-end (ray.timeline) to get profiling events from TaskEvents in GCS. 

**In Future PRs:**
1. Remove legacy code for profiling events 
2. State API porting. 

## Checks

```
RAY_PROFILING=1 pytest python/ray/tests/test_advanced.py::test_profiling_api -vs
```
image

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -461,17 +461,19 @@ RAY_CONFIG(int64_t, metrics_report_batch_size, 100)
 /// The interval duration for which task state events will be reported to GCS.
 /// The reported data should only be used for observability.
 /// Setting the value to 0 disables the task event recording and reporting.
-RAY_CONFIG(int64_t, task_events_report_interval_ms, 0)
+RAY_CONFIG(int64_t, task_events_report_interval_ms, 1000)
```
Comment (rickyyx) [inline src/ray/common/ray_config_def.h:464]: This turns the backend on by default - or it could be made into another independent PR.

Comment (rkooo567) [reply src/ray/common/ray_config_def.h:464]: I think it's better we turn this on by default asap (so that we can have more time to fix by 2.3 release). cc @ericl in case you disagree.

Comment (rickyyx) [reply src/ray/common/ray_config_def.h]: Oh actually, we will need to turn the feature on with this change, so that profiling timeline events will be available. 

We could run nightly on this PR to get a baseline performance number.

**Review by ericl — APPROVED** [review]

File: src/ray/common/ray_config_def.h [hunk 2]
```diff
@@ -465,13 +465,15 @@ RAY_CONFIG(int64_t, task_events_report_interval_ms, 0)
...
+/// cost less than 250MiB.
 /// Setting the value to -1 allows for unlimited task events stored in GCS.
-RAY_CONFIG(int64_t, task_events_max_num_task_in_gcs, 100000)
+RAY_CONFIG(int64_t, task_events_max_num_task_in_gcs, 500000)
 
 /// Max number of task events stored in the buffer on workers. Any additional events
 /// will be dropped.
 /// Setting the value to -1 allows for unlimited task events buffered on workers.
-RAY_CONFIG(int64_t, task_events_max_num_task_events_in_buffer, 10000)
+RAY_CONFIG(int64_t, task_events_max_num_task_events_in_buffer, 50000)
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:476]: Are these changes necessary?

**Comment by rickyyx** [comment]
Some selected benchmark results from the run: https://buildkite.com/ray-project/release-tests-pr/builds/24015#_

**tldr: A few microbenchmark metrics have some regression from up to 15%, while most of the e2e tests see a negligible impact.** 

| Metric                      | Master | PR    |
|-----------------------------|--------|-------|
| microbenchmark | | | 
| single_client_tasks_sync    | 1252   | 1238  |
| single_client_tasks_async   | 11180  | 10648 |
| multi_client_tasks_async    | 33785  | 28934 |
| 1_1_actor_calls_sync        | 2155   | 1881  |
| 1_1_actor_calls_async       | 7061   | 6926  |
| 1_n_actor_calls_async       | 10689  | 10822 |
| n_n_actor_calls_async       | 34513  | 30988 |
| n_n_async_actor_calls_async | 28600  | 25367 |
| | |
| stress_many_tasks | | 
| stage_0_time | 8.67 |  7.70  | 
| stage_1_avg_iteration_time | 22.27 | 22.87 | 
| stage_2_avg_iteration_time | 55.53 | 51.76| 
| stage_3_creation_time | 0.052 | 0.053 |
| | |
| many_tasks | | 
| tasks_per_second | 27.11 | 26.67 | 
| used_cpus_by_deadline | 1759 | 1753| 
| | |
| many_actors | | 
| actors_per_second | 292.82  | 435.88 | 
| | |
| single_node | | 
| args_time | 17.02 |  17.06 | 
| returns_time | 6.11 | 6.38 | 
| get_time | 25.07 | 25.47 | 
| queued_time |  198 | 205 |
| large_object_time | 259 |  251 | 
| | |
| many_nodes | | 
| tasks_per_second | 9.38|  9.15 |

**Comment by rickyyx** [comment]
Test failures unrelated (some python dependency issue + memory_monitor_test that's still flaky)

**Comment by ericl** [comment]
LGTM
